File: 20_Arrays/020_Reverse_Array.py
# Simple Swapping Method
def reverse_entire_list(array):
    start, end = 0, len(array) - 1  # Initialize 'start' and 'end' indices

    # Loop until 'start' index is less than 'end' index
    while start < end:
        array[start], array[end] = array[end], array[start]  # Swap elements
        start += 1  # Move 'start' index towards the center
        end -= 1  # Move 'end' index towards the center

# Main driver code
array = [1, 2, 3, 4, 5, 6]  # Original array
print("Simple original example array is:", array)

# Call 'reverse_entire_list' to reverse the entire array
reverse_entire_list(array)

# Print the result after reversal
print("Reversed of simple array is:", array)


# Stack Method
def reverse_array_using_stack(arr):
    stack = []  # Initialize an empty stack
    reversed_arr = []  # Initialize an empty array for the reversed elements

    # Push all elements of the input array onto the stack
    for element in arr:
        stack.append(element)  # Push element onto stack

    # Pop all elements from the stack to form the reversed array
    while stack:
        top_element = stack.pop()  # Pop the top element from the stack
        reversed_arr.append(top_element)  # Append the popped element to the reversed array

    return reversed_arr

# Main driver code
original_array = [1, 2, 3, 4, 5]
print("Stack Original array:", original_array)

# Call 'reverse_array_using_stack' to reverse the array and get a new reversed array
reversed_array = reverse_array_using_stack(original_array)

# Print the reversed array
print("Stack Reversed array using stack:", reversed_array)

File: 20_Arrays/020_Reverse_Array.cpp
#include <iostream>
#include <vector>
#include <stack>

// Function to print the contents of an array
void printArray(const std::vector<int>& array, const std::string& message) {
    std::cout << message;
    for (int num : array) {
        std::cout << num << " ";
    }
    std::cout << std::endl;
}

// Function to reverse the entire list using simple swapping
void reverseEntireList(std::vector<int>& array) {
    int start = 0; // Initialize 'start' index to the beginning of the array
    int end = array.size() - 1; // Initialize 'end' index to the end of the array

    while (start < end) {
        std::swap(array[start], array[end]);
        start++; // Move 'start' index towards the center
        end--; // Move 'end' index towards the center
    }
}

// Function to reverse an array using a stack
std::vector<int> reverseArrayUsingStack(const std::vector<int>& arr) {
    std::stack<int> stack; // Initialize an empty stack
    std::vector<int> reversedArr; // Initialize an empty vector for the reversed elements

    for (int element : arr) {
        stack.push(element); // Push element onto stack
    }

    while (!stack.empty()) {
        reversedArr.push_back(stack.top()); // Append the top element from the stack
        stack.pop(); // Remove the top element from the stack
    }

    return reversedArr;
}

int main() {
    // Demonstrate Simple Swapping Method
    std::vector<int> array = {1, 2, 3, 4, 5, 6};
    printArray(array, "Original array (Simple Swap): ");
    reverseEntireList(array);
    printArray(array, "Reversed array (Simple Swap): ");

    // Demonstrate Stack Method
    std::vector<int> originalArray = {1, 2, 3, 4, 5};
    printArray(originalArray, "Original array (Stack Method): ");
    std::vector<int> reversedArray = reverseArrayUsingStack(originalArray);
    printArray(reversedArray, "Reversed array (Stack Method): ");

    return 0;
}
